lib/charms/opensearch/v0/opensearch_data.py

```diff
@@ -0,0 +1,503 @@
+        if not self._jujuversion:
+            self._jujuversion = JujuVersion.from_environ()
+        if self._jujuversion and hasattr(self._jujuversion, "has_secrets"):
+            return self._jujuversion.has_secrets
+
+    def is_secret(self, scope: Scope, key: str) -> bool:
+        """Does 'key' correspond to a secret data within 'scope'?"""
+        if self.has_secrets:
+            return self.label(scope, key) in self.secrets
+        else:
+            # Pretty fragile, to be gone with Juju 2.9
+            if key in CertType.all_values():
+                return True
+            elif "-password" in key:
+                return True
+        return False
+
+    @staticmethod
+    def elapsed(f: callable, *args, **kwargs):
+        """Temporary time measurement method."""
+        start = time.time()
+        result = f(*args, **kwargs)
+        end = time.time()
+        logging.info(f"Secret operation {f} takes {end-start}")
+        return result
+
+    def _scope_obj(self, scope):
+        if scope == Scope.APP:
+            return self._charm.app
+        if scope == Scope.UNIT:
+            return self._charm.unit
+
+    def label(self, scope: Scope, key: str) -> str:
+        """Generated keys to be used within relation data to refer to secret IDs."""
+        components = [scope.val, self._scope_obj(scope).name, key]
+        return SECRET_LABEL_SEPARATOR.join(components)
+
+    @staticmethod
+    def _safe_data(inval: Union[bool, int, float, str]) -> str:
+        if not isinstance(inval, str):
+            return str(inval)
+        return inval
+
+    @staticmethod
+    def _safe_obj_data(indict: Dict) -> Dict:
+        outdict = copy.deepcopy(indict)
+
+        # https://stackoverflow.com/questions/11941817/how-can-i-avoid-runtimeerror-dictionary-changed-size-during-iteration-error
+        for k in list(outdict):
+            if outdict[k] is None:
+                outdict.pop(k)
+            else:
+                safe_val = SecretsDataStore._safe_data(outdict[k])
+                if safe_val != outdict[k]:
+                    outdict[k] = safe_val
+        return outdict
+
+    def _get_juju_secret(self, scope: Scope, key: str) -> Optional[Secret]:
+        label = self.label(scope, key)
+
+        cached_secret_meta = self.cached_secrets.get_meta(scope, label)
+        if cached_secret_meta:
+            return cached_secret_meta
+
+        try:
+            secret = self.elapsed(self._charm.model.get_secret, label=label)
+        except SecretNotFoundError:
+            return
+
+        self.cached_secrets.set_meta(scope, label, secret)
+        return secret
+
+    def _get_juju_secret_content(self, scope: Scope, key: str) -> Optional[Dict]:
+        cached_secret_contents = self.cached_secrets.get_contents(scope, self.label(scope, key))
+        if cached_secret_contents:
+            return cached_secret_contents
+
+        secret = self._get_juju_secret(scope, key)
+        if not secret:
+            return
```

Comment:
```suggestion
            return None
```

Reply:
Is it preferred to return `None` explictily?
If yes, I'd adjust ALL occurances (as otherwise it's kinda inconsistent).